README.rst
.. image:: https://img.shields.io/pypi/v/nengo-bones.svg
  :target: https://pypi.org/project/nengo-bones
  :alt: Latest PyPI version

.. image:: https://img.shields.io/codecov/c/github/nengo/nengo-bones/master.svg
  :target: https://codecov.io/gh/nengo/nengo-bones
  :alt: Test coverage

**********
NengoBones
**********

The aim of this project is to automate the configuration of repositories
throughout the Nengo ecosystem.  The idea is that we can centralize the
design and maintenance of "meta" project code (such as CI testing
infrastructure) in this project, rather than each repository maintaining that
code independently.

The basic methodology for this project is a templating system in which
there are common templates for meta files that are populated with
data in each downstream project.  Projects control this templating through the
``.nengobones.yml`` configuration file, which defines the information used to
fill in the templates.

Wherever possible, we try to do this templating in such a way that downstream
projects will be automatically updated when an update is made in
``nengo-bones``. However, some files cannot be updated automatically and
require downstream repos to manually run a script to update those files.

Note that this repository itself is configured using the ``nengo-bones``
templating system, so if you would like an example of how to use it, check out
`the source code <https://github.com/nengo/nengo-bones>`__.

Installation
============

NengoBones is designed to be used based on the ``master`` branch (although we
release occasional stable checkpoints to PyPI). This can be installed via

.. code-block:: bash

    git clone https://github.com/nengo/nengo-bones.git
    pip install -e ./nengo-bones

Basic usage
===========

The first step is to fill in the ``.nengobones.yml`` configuration file.  You
can use the one in this repository as a starting point, or see
`the documentation
<https://www.nengo.ai/nengo-bones/examples/configuration.html>`__
for more details.  This file should be
placed in the top level of your project.

All of the manually generated template files can then be rendered by running
this command in the same folder as the ``.nengobones.yml`` file:

.. code-block:: bash

    bones-generate

See ``bones-generate --help`` or
`the documentation <https://www.nengo.ai/nengo-bones/cli.html>`__
for a full list of command line options.

Documentation
=============

- `Command line usage <https://www.nengo.ai/nengo-bones/cli.html>`_
- `Demonstration of available configuration options
  <https://www.nengo.ai/nengo-bones/examples/configuration.html>`_
- `API reference <https://www.nengo.ai/nengo-bones/reference.html>`_


CHANGES.rst
***************
Release History
***************

.. Changelog entries should follow this format:

   version (release date)
   ======================

   **section**

   - One-line description of change (link to Github issue/PR)

.. Changes should be organized in one of several sections:

   - Added
   - Changed
   - Deprecated
   - Removed
   - Fixed

20.5 (unreleased)
=================

**Added**

- Added support for new ``one_page`` option in ``nengo-sphinx-theme``. (`#101`_)
- Added ``tagmanager_id`` option to ``docs_conf.py.template``,
  which will enable Google Tag Manager tracking. This option takes precedence
  over ``analytics_id`` if both are specified. (`#114`_)
- Added the ``bones-format-notebook`` script to apply automated formatting, cleanup,
  and static checking to Jupyter notebooks. (`#32`_)
- Static script will now check that ``bones-format-notebook`` has been applied to all
  notebooks in the ``docs`` directory. (`#32`_)
- Added ``remote-script.sh`` CI script for running other CI scripts on a remote device.
  (`#124`_)
- Added `isort <https://pycqa.github.io/isort/>`_ configuration to
  ``pyproject.toml`` and check import order during static checks. (`#121`_)
- Officially support and test against Python 3.9. (`#136`_)
- Added ``skip`` option to codespell config, which can be used to specify files that
  should be ignored. (`#138`_)
- Added support for projects with type hints through the ``py_typed`` section. (`#140`_)
- Added ``min_python`` option to the main section. The default ``python_requires``
  in ``setup.py`` is now based on this, if not overridden. (`#140`_)
- Added support for changing the main branch name with the ``main_branch``
  config option. (`#145`_)
- Added checksum verification to Codecov upload script. (`#150`_)

**Changed**

- Restrict pylint version to <2.5 due to issue when the package being checked
  is not installed in the current environment. (`#103`_)
- Increased minimum ``black`` version to 20.8b0. (`#104`_)
- Increased minimum ``pylint`` version to 2.5.1 (`#32`_)
- Documentation ``[source]`` links now point to GitHub. (`#117`_)
- Separate tokens, ``PYPI_TOKEN`` and ``PYPI_TEST_TOKEN``, must now be specified
  for deploying to PyPI and TestPyPI, respectively. (`#127`_)
- Builds for this repository now run on TravisCI.com instead of TravisCI.org. (`#130`_)
- Drop support for Python 3.5. (`#123`_)
- Doc script will now install Sphinx>=3.1.2. (`#137`_)
- Codespell ignore words is now specified via the ``codespell.ignore_words`` section
  of ``setup_cfg`` (instead of in the static CI script config). As a result, it will
  now apply to all invocations of codespell (not just from the static script). (`#138`_)
- ``bones-format-notebook --check`` will now require that all ``codespell`` checks pass.
  (`#138`_)

**Fixed**

- Intermittent failures installing miniconda in the remote script have been fixed by
  retrying the miniconda install if it fails. (`#130`_)
- Fixed an issue in which Codecov would not report on files where a file in a
  deeper folder shares the same name. (`#132`_)
- Fixed an issue in which CI scripts did not have access to the same configuration
  options as other templates. (`#144`_)

.. _#32: https://github.com/nengo/nengo-bones/pull/32
.. _#101: https://github.com/nengo/nengo-bones/pull/101
.. _#103: https://github.com/nengo/nengo-bones/pull/103
.. _#104: https://github.com/nengo/nengo-bones/pull/104
.. _#114: https://github.com/nengo/nengo-bones/pull/114
.. _#117: https://github.com/nengo/nengo-bones/pull/117
.. _#121: https://github.com/nengo/nengo-bones/pull/121
.. _#123: https://github.com/nengo/nengo-bones/pull/123
.. _#124: https://github.com/nengo/nengo-bones/pull/124
.. _#127: https://github.com/nengo/nengo-bones/pull/127
.. _#130: https://github.com/nengo/nengo-bones/pull/130
.. _#132: https://github.com/nengo/nengo-bones/pull/132
.. _#136: https://github.com/nengo/nengo-bones/pull/136
.. _#137: https://github.com/nengo/nengo-bones/pull/137
.. _#138: https://github.com/nengo/nengo-bones/pull/138
.. _#140: https://github.com/nengo/nengo-bones/pull/140
.. _#144: https://github.com/nengo/nengo-bones/pull/144
.. _#145: https://github.com/nengo/nengo-bones/pull/145
.. _#150: https://github.com/nengo/nengo-bones/pull/150

0.11.1 (April 13, 2020)
=======================

**Changed**

- Rendered documentation will not be uploaded if the html build fails (it will still
  be uploaded if the linkchecker/doctest builds fail). (`#98`_)
- Rendered documentation will not be uploaded on cron builds. (`#98`_)
- Docs script will now clean up the built doc directory before execution, if it exists
  (e.g., because the docs job is being rerun). (`#96`_)

.. _#96: https://github.com/nengo/nengo-bones/pull/96
.. _#98: https://github.com/nengo/nengo-bones/pull/98

0.11.0 (April 13, 2020)
=======================

**Added**

- Downstream repos will now be automatically updated when nengo-bones is updated.
  (`#97`_)
- Added ``slack_notifications`` option to ``.travis.yml`` to enable Slack notifications
  for failing builds. (`#97`_)

**Changed**

- Will now use ``nengo-bones`` and ``nengo-sphinx-theme`` master builds (instead of the
  latest release), to streamline the process of distributing changes to those core
  repos. (`#97`_)

.. _#97: https://github.com/nengo/nengo-bones/pull/97

0.10.0 (March 19, 2020)
=======================

**Added**

- Added ``autoautosummary_change_modules`` option (for use with
  ``nengo_sphinx_theme.ext.autoautosummary``). (`#86`_)

**Changed**

- Docs script will now use ``nengo_sphinx_theme.ext.backoff``, which adds
  exponential backoff functionality to Sphinx requests. (`#86`_)

.. _#86: https://github.com/nengo/nengo-bones/pull/86

0.9.1 (March 17, 2020)
======================

**Fixed**

- Fixed deployment tag conditional check in ``.travis.yml`` template. (`#83`_)

.. _#83: https://github.com/nengo/nengo-bones/pull/83


0.9.0 (January 28, 2020)
========================

**Changed**


nengo_bones/tests/test_config.py
# pylint: disable=missing-docstring

import os

import pytest

from nengo_bones import config
from nengo_bones.tests import utils


def test_find_config():
    assert config.find_config() == os.path.join(os.getcwd(), ".nengobones.yml")


def test_fill_defaults():
    init_cfg = {
        "travis_yml": {"jobs": [{"script": "docs-test"}]},
        "codecov_yml": {},
        "version": {
            "major": 3, "minor": 2, "patch": 1, "release": False,
        }
    }
    config.fill_defaults(init_cfg)

    assert init_cfg["travis_yml"]["python"] == "3.6"
    assert init_cfg["travis_yml"]["global_vars"] == {}
    assert init_cfg["travis_yml"]["pypi_user"] is None
    assert init_cfg["travis_yml"]["deploy_dists"] == ["sdist"]

    assert init_cfg["travis_yml"]["jobs"][0]["apt_install"] == ["pandoc"]

    assert init_cfg["codecov_yml"]["skip_appveyor"]
    assert init_cfg["codecov_yml"]["abs_target"] == "auto"
    assert init_cfg["codecov_yml"]["diff_target"] == "100%"

    assert init_cfg["version"]["full"] == "3.2.1.dev0"


def test_validate_config():
    mandatory = [
        "project_name", "pkg_name", "repo_name", "travis_yml.jobs",
        "version", "version.major", "version.minor", "version.patch",
        "version.release",
    ]
    init_cfg = {"travis_yml": {}}
    for entry in mandatory:
        with pytest.raises(KeyError, match="must define %s" % entry):
            config.validate_config(init_cfg)

        keys = entry.split(".")
        tmp = init_cfg
        for key in keys[:-1]:
            tmp = tmp[key]
        tmp[keys[-1]] = {}

    config.validate_config(init_cfg)

    init_cfg["ci_scripts"] = {}

    # error when template not defined
    test_cfg = {}
    init_cfg["ci_scripts"] = [test_cfg]
    with pytest.raises(KeyError, match="must define 'template'"):
        config.validate_config(init_cfg)
    test_cfg["template"] = "test"

    # error when pip_install is a string instead of list
    test_cfg["pip_install"] = "pip_req"
    with pytest.raises(TypeError, match="pip_install should be a list"):
        config.validate_config(init_cfg)
    test_cfg["pip_install"] = ["pip_req"]

    # error when pre_commands is a string instead of list
    test_cfg["pre_commands"] = "command"
    with pytest.raises(TypeError, match="pre_commands should be a list"):
        config.validate_config(init_cfg)


def test_missing_config(tmpdir):
    with pytest.raises(RuntimeError, match="Could not find conf_file"):
        config.load_config(tmpdir.join(".does-not-exist.yml"))


def test_load_config(tmpdir):
    truth = {
        "project_name": "Dummy",
        "pkg_name": "dummy",
        "repo_name": "dummyorg/dummy",
        "version": {
            "major": 3,
            "minor": 2,
            "patch": 1,
            "release": False,
            "full": "3.2.1.dev0",
        },
        "author": "A Dummy",
        "author_email": "dummy@dummy.com",
        "copyright_start": 0,
        "copyright_end": 1,
        "ci_scripts": [
            {"template": "static",
             "pip_install": ["static_pip0", "static_pip1"]}
        ],
        "travis_yml": {
            "python": "6.0",
            "global_vars": {
                "TEST_VAR": "test var val"},
            "pypi_user": "dummy_pypi_user",
            "deploy_dists": [
                "sdist",
                "bdist_wheel"
            ],
            "jobs": [{"script": "static", "language": "generic"}],
            "bones_install": "nengo-bones",
        },
        "codecov_yml": {
            "skip_appveyor": False,
            "abs_target": "test_abs",
            "diff_target": "test_diff"
        },
    }

    utils.write_file(tmpdir, ".nengobones.yml", """
        project_name: Dummy
        pkg_name: dummy
        repo_name: dummyorg/dummy
        version:
          major: 3
          minor: 2
          patch: 1
          release: false
        author: A Dummy
        author_email: dummy@dummy.com
        copyright_start: 0
        copyright_end: 1

        ci_scripts:
          - template: static
            pip_install:
              - static_pip0
              - static_pip1

        travis_yml:
          python: "6.0"
          global_vars:
            TEST_VAR: test var val
          pypi_user: dummy_pypi_user
          deploy_dists:
            - sdist
            - bdist_wheel
          jobs:
            - script: static
              language: generic

        codecov_yml:
          skip_appveyor: false
          abs_target: test_abs
          diff_target: test_diff
        """)

    loaded = config.load_config(tmpdir.join(".nengobones.yml"))

    try:
        assert loaded == truth
    except AssertionError:  # pragma: no cover
        print("loaded")
        print("\n".join("%s: %s" % (k, v) for k, v in loaded.items()))
        print("truth")
        print("\n".join("%s: %s" % (k, v) for k, v in truth.items()))
        raise

    # check loading from cwd
    cwd = os.getcwd()
    os.chdir(str(tmpdir))
    try:
        loaded = config.load_config()
    finally:
        os.chdir(cwd)

    assert loaded == truth


nengo_bones/tests/test_pr_number.py
# pylint: disable=missing-docstring

from functools import partial
import os

from click.testing import CliRunner
import pytest

from nengo_bones.scripts import pr_number
from nengo_bones.tests.utils import assert_exit, make_has_line, write_file


def mocked_requests_get(url, repo=None, number=None):
    """Used to mock out github requests so that we don't get timed out."""
    class MockResponse:
        def __init__(self, json_data):
            self.json_data = json_data

        def json(self):
            return self.json_data

    assert url == (
        "https://api.github.com/repos/%s/issues?state=all&sort=created" % repo)

    return MockResponse([{"number": number}])


@pytest.mark.parametrize("mode", ("repo", "conf-default", "conf-arg"))
def test_pr_number(monkeypatch, tmpdir, mode):
    repo = "a-repo/%s" % mode

    monkeypatch.setattr(
        pr_number.requests, "get",
        partial(mocked_requests_get, repo=repo, number=len(mode)))

    if mode != "repo":
        write_file(tmpdir, ".nengobones.yml", """
            project_name: Nengo Bones
            pkg_name: nengo-bones
            repo_name: %s
            version:
              major: 3
              minor: 2
              patch: 1
              release: False
        """ % repo)

    if mode == "repo":
        result = CliRunner().invoke(
            pr_number.main,
            [repo])
    elif mode == "conf-arg":
        result = CliRunner().invoke(
            pr_number.main,
            ['--conf-file',
             str(tmpdir.join('.nengobones.yml'))])
    else:
        original_dir = os.getcwd()
        os.chdir(str(tmpdir))
        try:
            result = CliRunner().invoke(pr_number.main)
        finally:
            os.chdir(original_dir)

    assert_exit(result, 0)

    lines = result.output.split('\n')
    has_line = make_has_line(lines, regex=True)
    assert has_line(r"about %s\.\.\." % repo)
    assert has_line(r"assigned #%s" % (len(mode) + 1))


nengo_bones/scripts/generate_bones.py
"""Scripts for auto-generating nengo-bones files."""

from collections import OrderedDict
import os
import stat

import click
import jinja2

import nengo_bones


def render_template(ctx, cfg_key, output_file, template_file=None, **kwargs):
    """
    Render a template to file.

    Parameters
    ----------
    ctx : `click.Context`
        CLI context, containing information specified upstream.
    cfg_key : str
        The heading for the section in the config file containing config
        options specific to the template being rendered.
    output_file : str
        Filename for the rendered output file.
    template_file
        Filename for the input template file (default is
        ``output_file + ".template"``)
    kwargs : dict
        Will be passed on to the ``render`` function.
    """
    if template_file is None:
        template_file = output_file + ".template"

    template = ctx.obj["env"].get_template(template_file)

    output_file = os.path.join(ctx.obj["output_dir"], output_file)
    with open(output_file, "w") as f:
        f.write(template.render(
            bones_version=nengo_bones.__version__,
            **kwargs,
            # pass in the top-level config options as well
            # TODO: separate "top-level" config into its own section?
            **ctx.obj["config"],
            **ctx.obj["config"][cfg_key],
        ))


@click.group(invoke_without_command=True)
@click.option("--conf-file", default=None, help="Filepath for config file")
@click.option("--output-dir", default=".", help="Output directory for scripts")
@click.pass_context
def main(ctx, conf_file, output_dir):
    """Loads config file and sets up template environment.

    By default, this updates all templated files that are
    to be committed to the repository.

    We look in the current directory for a ``.templates`` folder. If it
    exists, any templates defined in that folder will be loaded first.
    Otherwise, built-in templates will be loaded from the
    ``nengo_bones/templates`` directory.

    If you are overriding a template in the ``.templates`` folder, the
    original built-in version of that template can be accessed with
    the ``templates/`` prefix. This is useful in ``include`` and
    ``extends`` tags. For example, to add text to the default
    ``LICENSE.rst`` template, put the following in
    ``.templates/LICENSE.rst.template``:

    .. code-block:: rst

       {% include "templates/LICENSE.rst.template %}

       Additional license info
       =======================
       ...

    """

    ctx.ensure_object(dict)

    config = nengo_bones.load_config(conf_file)

    bones_toplevel = os.path.normpath(
        os.path.join(os.path.dirname(__file__), "..")
    )
    # Load overridden templates first.
    # Builtins are referenced with templates/*.template
    override_dirs = []
    override_dirs.append(".templates")
    override_dirs.append(bones_toplevel)
    override_loader = jinja2.FileSystemLoader(override_dirs)
    # If those fail, use the builtins
    builtin_loader = jinja2.FileSystemLoader(
        os.path.join(bones_toplevel, "templates")
    )

    env = jinja2.Environment(
        loader=jinja2.ChoiceLoader([override_loader, builtin_loader]),
        trim_blocks=True, lstrip_blocks=True, keep_trailing_newline=True)
    env.filters["rstrip"] = lambda s, chars: s.rstrip(chars)

    if not os.path.exists(output_dir):
        os.makedirs(os.path.normpath(output_dir))

    ctx.obj["env"] = env
    ctx.obj["config"] = config
    ctx.obj["output_dir"] = output_dir

    def check_cfg(name):
        name = name.replace("-", "_")
        if name not in config:
            print("No config entry detected for %s, skipping" % name)
            return False
        return True

    if ctx.invoked_subcommand is None:
        for cfg_name in nengo_bones.all_templated_files.values():
            if check_cfg(cfg_name):
                ctx.invoke(globals()[cfg_name])
    elif not check_cfg(ctx.invoked_subcommand):
        exit()


@main.command()
@click.pass_context
def ci_scripts(ctx):
    """Generate TravisCI shell scripts."""

    config = ctx.obj["config"]
    for params in config["ci_scripts"]:
        script_name = params.pop("template")
        template = ctx.obj["env"].get_template(script_name + ".sh.template")
        output_name = params.pop("output_name", script_name)
        path = os.path.join(ctx.obj["output_dir"], output_name + ".sh")

        with open(path, "w") as f:
            f.write(template.render(
                pkg_name=config["pkg_name"], repo_name=config["repo_name"],
                bones_version=nengo_bones.__version__, **params))

        # Mark CI script as executable
        st = os.stat(path)
        os.chmod(path, st.st_mode | stat.S_IEXEC)


@main.command()
@click.pass_context
def travis_yml(ctx):
    """Generate TravisCI config file."""

    config = ctx.obj["config"]

    jobs = config["travis_yml"]["jobs"]
    for job in jobs:
        # shortcuts for setting environment variables
        if "env" not in job:
            job["env"] = OrderedDict()
        for var in ("script", "test_args"):
            if var in job:
                job["env"][var] = job.pop(var)

    template = ctx.obj["env"].get_template(".travis.yml.template")
    with open(os.path.join(ctx.obj["output_dir"], ".travis.yml"), "w") as f:
        f.write(template.render(bones_version=nengo_bones.__version__,
                                **config["travis_yml"]))


@main.command()
@click.pass_context
def codecov_yml(ctx):
    """Generate codecov config file."""

    render_template(ctx, "codecov_yml", ".codecov.yml")


@main.command()
@click.pass_context
def license_rst(ctx):
    """Generate LICENSE.rst file."""

    render_template(ctx, "license_rst", "LICENSE.rst")


@main.command()
@click.pass_context
def contributing_rst(ctx):
    """Generate CONTRIBUTING.rst file."""

    render_template(ctx, "contributing_rst", "CONTRIBUTING.rst")


@main.command()
@click.pass_context
def contributors_rst(ctx):
    """Generate CONTRIBUTORS.rst file."""

    render_template(ctx, "contributors_rst", "CONTRIBUTORS.rst")


@main.command()
@click.pass_context
def manifest_in(ctx):
    """Generate MANIFEST.in file."""

    render_template(ctx, "manifest_in", "MANIFEST.in",
                    custom=ctx.obj["config"]["manifest_in"])


@main.command()
@click.pass_context
def setup_py(ctx):
    """Generate setup.py file."""

    config = ctx.obj["config"]

    setup_config = config["setup_py"]

    extras = OrderedDict()
    extra_types = {
        "classifiers": "list",
        "entry_points": "dict",
        "package_data": "dict",
    }
    # we iterate over setup_config (rather than extra_types) so that the
    # order in setup_config is preserved
    for key in list(setup_config.keys()):
        if key in extra_types:
            extras[key] = (extra_types[key], setup_config.pop(key))

    render_template(ctx, "setup_py", "setup.py",
                    extras=extras)


@main.command()
@click.pass_context
def setup_cfg(ctx):
    """Generate setup.cfg file."""

    render_template(ctx, "setup_cfg", "setup.cfg")


@main.command()
@click.pass_context
def docs_conf_py(ctx):
    """Generate docs/conf.py file."""

    output_dir = os.path.join(ctx.obj["output_dir"], "docs")

    if not os.path.exists(output_dir):
        os.makedirs(output_dir)

    render_template(ctx, "docs_conf_py", "docs/conf.py",
                    template_file="docs_conf.py.template")


@main.command()
@click.pass_context
def gitignore(ctx):
    """Generate setup.cfg file."""

    render_template(ctx, "gitignore", ".gitignore")


@main.command()
@click.pass_context
def version_py(ctx):
    """Generate {{ pkg_name }}/version.py file."""

    pkg_name = ctx.obj["config"]["pkg_name"]
    output_dir = os.path.join(ctx.obj["output_dir"], pkg_name)

    if not os.path.exists(output_dir):
        os.makedirs(output_dir)

    render_template(ctx, "version_py", "%s/version.py" % (pkg_name,),
                    template_file="version.py.template")


if __name__ == "__main__":
    main(obj={})  # pragma: no cover pylint: disable=no-value-for-parameter


nengo_bones/config.py
"""Handles the processing of nengo-bones configuration settings."""

from collections import OrderedDict
import datetime
import os

import yaml


def check_list(cfg, key):
    """
    Verify that config value is a list.

    This is designed to catch the common error of specifying

    .. code-block:: yaml

        option:
            value

    instead of

    .. code-block:: yaml

        option:
            - value

    Parameters
    ----------
    cfg : dict
        Configuration options being checked.
    key: str
        Name of configuration value to be checked.

    Raises
    ------
    TypeError
        If ``key`` is in ``cfg`` and its value is not a list.
    """
    if key in cfg and not isinstance(cfg[key], list):
        raise TypeError(
            "%s should be a list, found '%s'; did you forget "
            "to add '-' before each entry?" % (key, cfg[key]))


def find_config():
    """
    Finds the default nengo-bones config file.

    Returns
    -------
    conf_file : str
        Path to the default config file.
    """
    # for now, assume that config file is in cwd
    conf_file = os.path.join(os.getcwd(), ".nengobones.yml")

    return conf_file


def fill_defaults(config):
    """
    Fills in default values in a loaded config (in-place).

    Parameters
    ----------
    config : dict
        Dictionary containing configuration values.
    """

    config.setdefault("author", "Applied Brain Research")
    config.setdefault("author_email", "info@appliedbrainresearch.com")
    config.setdefault("copyright_start", datetime.datetime.now().year)
    config.setdefault("copyright_end", datetime.datetime.now().year)

    if "travis_yml" in config:
        cfg = config["travis_yml"]
        cfg.setdefault("python", "3.6")
        cfg.setdefault("global_vars", OrderedDict())
        cfg.setdefault("pypi_user", None)
        cfg.setdefault("deploy_dists", ["sdist"])
        cfg.setdefault("bones_install", "nengo-bones")

        for job in cfg["jobs"]:
            if job.get("script", "").startswith("docs"):
                job.setdefault("apt_install", ["pandoc"])

    if "codecov_yml" in config:
        cfg = config["codecov_yml"]
        cfg.setdefault("skip_appveyor", True)
        cfg.setdefault("abs_target", "auto")
        cfg.setdefault("diff_target", "100%")

    if "license_rst" in config:
        cfg = config["license_rst"]
        cfg.setdefault("type", "nengo")

    if "setup_py" in config:
        cfg = config["setup_py"]
        cfg.setdefault("license", "Free for non-commercial use")
        cfg.setdefault("python_requires", ">=3.5")
        cfg.setdefault("include_package_data", False)
        cfg.setdefault(
            "url",
            "https://www.nengo.ai/%s" % config["pkg_name"].replace("_", "-"))

    if "setup_cfg" in config:
        cfg = config["setup_cfg"]
        cfg.setdefault("pytest", OrderedDict())
        cfg.setdefault("pylint", OrderedDict())
        cfg.setdefault("flake8", OrderedDict())
        cfg.setdefault("coverage", OrderedDict())
        cfg["pytest"].setdefault("xfail_strict", False)
        cfg["pytest"].setdefault("addopts", ["-p nengo.tests.options"])

    if "docs_conf_py" in config:
        cfg = config["docs_conf_py"]
        cfg.setdefault("nengo_logo", "general-full-light.svg")

    # Automatically fill in full version
    v = config["version"]
    v["full"] = "{}.{}.{}{}".format(
        v["major"], v["minor"], v["patch"], "" if v["release"] else ".dev0"
    )


def validate_config(config):
    """
    Validates a populated config dict.

    Parameters
    ----------
    config : dict
        Dictionary containing configuration values.
    """
    mandatory = [
        "project_name", "pkg_name", "repo_name", "travis_yml.jobs",
        "version", "version.major", "version.minor", "version.patch",
        "version.release",
    ]

    for entry in mandatory:
        tmp = config
        for i, key in enumerate(entry.split(".")):
            try:
                tmp = tmp[key]
            except KeyError:
                if "." in entry and i == 0:
                    # if the toplevel isn't defined, ignore this
                    break
                else:
                    raise KeyError("Config file must define %s" % entry)

    if "ci_scripts" in config:
        for ci_config in config["ci_scripts"]:
            validate_ci_config(ci_config)

    if "gitignore" in config:
        check_list(config["gitignore"], "extra")

    # TODO: check that there aren't unused config options in yml


def validate_ci_config(ci_config):
    """
    Validates an entry in the ci_scripts list of a config dict.

    Parameters
    ----------
    ci_config : dict
        Dictionary containing ci_scripts configuration values.
    """
    if "template" not in ci_config:
        raise KeyError("Script config must define 'template' "
                       "(for entry %s)" % ci_config)

    # make sure that people don't accidentally do things like
    # pip_install: dependency (which gives a string), rather than
    # pip_install:
    #   - dependency
    list_opts = ("pip_install", "pre_commands", "post_commands",
                 "codespell_ignore_words")
    for opt in list_opts:
        check_list(ci_config, opt)


def load_config(conf_file=None):
    """
    Loads config values from a file and applies defaults/validation.

    Parameters
    ----------
    conf_file : str
        Filepath for config file (if None, will load the default returned by
        `.find_config`).

    Returns
    -------
    config : dict
        Dictionary containing configuration values.
    """

    if conf_file is None:
        conf_file = find_config()

    if not os.path.exists(str(conf_file)):
        raise RuntimeError("Could not find conf_file: %s\n\nPerhaps you are "
                           "not in the project's root directory?" % conf_file)

    def ordered_load(stream):
        """Use OrderedDict instead of dict for loading mappings."""

        class OrderedLoader(yaml.SafeLoader):  # pylint: disable=too-many-ancestors
            """Custom loader containing OrderedDict mapping."""

        def construct_mapping(loader, node):
            loader.flatten_mapping(node)
            return OrderedDict(loader.construct_pairs(node))

        OrderedLoader.add_constructor(
            yaml.resolver.BaseResolver.DEFAULT_MAPPING_TAG,
            construct_mapping)
        return yaml.load(stream, OrderedLoader)

    with open(str(conf_file)) as f:
        config = ordered_load(f)

    validate_config(config)

    fill_defaults(config)

    return config
